pypeman/tests/test_msgstore.py

```diff
@@ -1,472 +1,518 @@
+
+    msg = Message(
+        payload="thingymabob",
+        meta={
+            "": "empty named meta",
+            "meta that is a list": [1, 2, 3],
+            "true": False,
+            "queneni": None,
+        },
+    )
+    id = await store.store(msg)
+    assert type(id) is str
+
+    entry = await store.get(id)
+    assert entry["id"] == id
+
+    remsg = entry["message"]
+    assert remsg.timestamp == msg.timestamp
+    assert remsg.uuid == msg.uuid
+    assert remsg.payload == msg.payload
+    assert remsg.meta == msg.meta
+    assert remsg.ctx == msg.ctx
+
+
+async def test_store_delete(factory: MessageStoreFactory):
+    "Using an `id` after a call to :meth:`MessageStore.delete` fails."
+    store = factory.get_store("a")
+    await store.start()
+
+    id = await store.store(Message())
+    _ = await store.get(id)  # store successful, message exists
+    await store.delete(id)
+
+    # do it with both implementation methods directly because we want
+    # to make sure both work (one might always shadow the other if eg the
+    # meta is always used first in implementation of says `.get`)
+    rose = None
+    try:
+        _ = await store._get_storemeta(id)
+    except BaseException as exc:
+        rose = exc
+    assert isinstance(rose, LookupError), rose
+    try:
+        _ = await store._get_message(id)
+    except BaseException as exc:
+        rose = exc
+    assert isinstance(rose, LookupError), rose
+
+
+async def test_get_invalid(factory: MessageStoreFactory):
+    "Using a wrong `id` fails."
+    store = factory.get_store("a")
+    await store.start()
+
+    id = await store.store(Message())
+
+    rose = None
+    try:
+        _ = await store._get_storemeta("nop!" + id)
+    except BaseException as exc:
+        rose = exc
+    assert isinstance(rose, LookupError), rose
+    try:
+        _ = await store._get_message("nop!" + id)
+    except BaseException as exc:
+        rose = exc
+    assert isinstance(rose, LookupError), rose
+
+
+async def test_store_state_management(factory: MessageStoreFactory):
+    "Updates to the state"
+    store = factory.get_store("a")
+    await store.start()
+
+    id = await store.store(Message())
+    assert await store.get_message_state(id) == Message.PENDING
+
+    await store.change_message_state(id, Message.PROCESSING)
+    assert await store.get_message_state(id) == Message.PROCESSING
+
+    await store.change_message_state(id, Message.PROCESSED)
+    assert await store.get_message_state(id) == Message.PROCESSED
+
+
+async def test_double_store_message(factory: MessageStoreFactory):
+    """By double store I mean twice the same, (mostly the same uuid).
+
+    The behavior in this situation is not really specified yet:
+        * Return id of the existing message without modifying it?
+        * Update/replace and return same or even new id?
+        * Raise an error (making it possible to find the bug)?
+
+    TODO: for now this is fake coverage! (really minor but still fake!)
```

Comment:
For traceability purpose, add a "replay_of" msg's meta with the uuid of the origin message.